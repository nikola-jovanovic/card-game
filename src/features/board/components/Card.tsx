import React, { memo } from 'react'
import styled from 'styled-components'

import { Card as CardT } from '../types'

export type Props = CardT & {
  facedown?: boolean
  onClick?: any
}

const Image = styled.img`
  max-width: 80px;
`

const Button = styled.button`
  border: none;
  padding: 0;
  margin-right: 4px;
  background-color: transparent;
  cursor: pointer;
`

const Card = ({
  code,
  value,
  image,
  facedown,
  onClick,
}: Props): JSX.Element => {
  if (facedown) {
    return (
      <Image
        src="https://dcassetcdn.com/design_img/3930521/154340/26190370/kfmqknarypcs90a4zrbf3ppavp_thumbnail.png"
        aria-label="facedown card"
      />
    )
  }

  const handleClick = () => onClick({ code, value, image })
  const img = <Image src={image} aria-label={code} />

  return onClick ? (
    <Button onClick={handleClick} aria-label={code}>
      {img}
    </Button>
  ) : img
}

export default memo(Card)
