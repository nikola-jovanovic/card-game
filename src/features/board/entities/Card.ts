import { chain } from 'fp-ts/Either'
import { pipe } from 'fp-ts/function'
import { exact,string, success, Type, type } from 'io-ts'

const Cards: { [index: string]: number } = {
  'ACE': 1,
  'JACK': 12,
  'QUEEN': 13,
  'KING': 14
}

const value = new Type<number, string, unknown>(
  'cardValue',
  (u): u is number => typeof u === 'number',
  (u, c) =>
    pipe(
      string.validate(u, c),
      chain((s: string) => {
        const d = parseInt(s)

        return success(isNaN(d) ? Cards[s] : d)
      }),
    ),
  (a) => a.toString(),
)

export const Card = exact(
  type(
    {
      code: string,
      image: string,
      value,
    },
    'Card',
  ),
)
