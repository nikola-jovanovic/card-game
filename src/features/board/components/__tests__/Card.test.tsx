import React from 'react'
import { render, screen, fireEvent } from '@testing-library/react'

import Card, { Props } from '../Card'

describe('Card', () => {
  let props: Props

  beforeAll(() => {
    props = {
      code: 'JH',
      image: 'image-url',
      value: 5
    }
  })

  it('render', () => {
    const onClick = jest.fn()
    const { container } = render(<Card {...props} onClick={onClick} />)

    const button = screen.getByRole('button', { name: /jh/i })

    expect(button).toBeInTheDocument()
    expect(screen.getByRole('img', { name: /jh/i })).toBeInTheDocument()

    expect(container).toMatchSnapshot()

    fireEvent.click(button)

    expect(onClick).toHaveBeenCalledTimes(1)
    expect(onClick).toHaveBeenCalledWith(props)
  })

  it('facedown', () => {
    const { container } = render(<Card {...props} facedown />)

    expect(screen.getByRole('img', { name: /facedown/i })).toBeInTheDocument()

    expect(container).toMatchSnapshot()
  })
})