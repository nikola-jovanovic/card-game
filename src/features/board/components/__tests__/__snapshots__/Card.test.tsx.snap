// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card facedown 1`] = `
.c0 {
  max-width: 80px;
}

<div>
  <img
    aria-label="facedown card"
    class="c0"
    src="https://dcassetcdn.com/design_img/3930521/154340/26190370/kfmqknarypcs90a4zrbf3ppavp_thumbnail.png"
  />
</div>
`;

exports[`Card render 1`] = `
.c1 {
  max-width: 80px;
}

.c0 {
  border: none;
  padding: 0;
  margin-right: 4px;
  background-color: transparent;
  cursor: pointer;
}

<div>
  <button
    aria-label="JH"
    class="c0"
  >
    <img
      aria-label="JH"
      class="c1"
      src="image-url"
    />
  </button>
</div>
`;
